oeps/oep-0031-arch-i18n.rst

```diff
@@ -0,0 +1,122 @@
+==============================
+
+.. list-table::
+
+   * - OEP
+     - :doc:`OEP-0031 <oep-0031-arch-i18n>`
+   * - Title
+     - Internationalization
+   * - Last Modified
+     - 2019-01-07
+   * - Authors
+     -
+   * - Arbiter
+     -
+   * - Status
+     - Draft
+   * - Type
+     - Architecture
+   * - Created
+     - 2019-01-07
+   * - Review Period
+     -
+
+.. contents::
+   :local:
+   :depth: 2
+
+Motivation
+==========
+
+Internationalization (i18n) is a vital part of unlocking content for users around the globe and effectively supporting our business partners. In order to help us reuse localized components from app to app, the different parts of the edX codebase should align on a consistent handling of i18n. This document covers our approach to internationalizing new front-end applications and components written in React.
+
+In this OEP we use the terms locale and language more or less interchangeably. However, note that this document addresses primarily making code multilingual, i.e., the issue of presenting the same content in different languages. Considering ways to make it multi-regional, to allow for presenting different content for different countries, is out of scope for this OEP.
+
+Technology Choice
+=================
+
+We will use `react-intl <https://github.com/yahoo/react-intl>`_ to internationalize React components, in agreement with `Paragon ADR-2 <https://github.com/edx/paragon/blob/f5b9f876800e5154ef01d04f682b8cda21b3e439/docs/decisions/0002-react-i18n.rst>`_. react-intl is in line with current industry standards in i18n and satisfies all currently identified i18n needs, including multiple plurals.
+
+Alternatives Considered
+-----------------------
+
+An alternative choice we considered was `i18next <https://react.i18next.com/>`_. **[TODO]**
+
+Locale
+======
+
+Localized content for new micro frontends should be at a URL that includes the language as the first part of the path after the domain. (Example: ``account.edx.org/es/``) The language string should be in the form of a `BCP 47 language tag <http://tools.ietf.org/html/rfc5646>`_. If the user enters the site at a URL that doesn't include a language string, we will determine a locale and then forward them to the corresponding URL. For consistency, this applies even if their language is English: ``account.edx.org/`` should be forwarded to ``account.edx.org/en/`` . We will use the most general, two-letter language codes to start with, and re-evaluate as necessary.
+
+When choosing what locale to use for a page, the locale in the URL always takes precedence. If there is none, check the user's cookies for a locale; if there isn't one, fetch it from the user's account settings and store it in their cookies for reuse. If there is no locale (or no locale we support) in the account settings, we will use the first locale we can support in the ``Accept-Language`` header that was sent with the request. Failing everything else, we will fall back to English.
```

Comment:
*nod*  @rock345  's team has a specific requirement to show a given locale regardless of the user's settings, which is how this came about.  Maybe there's a better way to do it?  Either way, we could certainly offer the user a small banner notification asking if they'd like to change to their preferred locale if we saw that they were on a URL that wasn't what was in their cookie.  That would cover the case of users getting a bookmark from a friend, as well.

Reply:
Not giving precedence to the user's own settings does seem confusing.  Can we reference a product decision document or such that explains the rationale for this decision?

Reply:
For the marketing site (or any site that support anonymous browsing) they should be shown /lang-code/ if that is the URL they have arrived at.  This should not redirect to /en or another language regardless of their user preference

Reply:
moving my comment into this thread now that I see it has been discussed already: 

"When choosing what locale to use for a page, the locale in the URL always takes precedence."

Thoughts here on how this impacts shared urls? If I share a url to an edX page that includes my locale and someone with a Spanish language clicks the link, it sounds like the url precedence rule would mean they would be switched over to English instead of being redirected to the Spanish version of the page (which is their default / account setting)

Reply:
@marcotuts Your last paragraph correctly describes the current proposal.  I am happy to give you and @rock345 giant q-tips and have you battle it out.  :-)

![image](https://user-images.githubusercontent.com/45690905/51689496-d17a1200-1fc4-11e9-8535-6cad1ed14d42.png)

Reply:
@albemarle I think everyone here is looking for the reason behind why the locale in the URL take precedence over user account setting value. If there is no good reason for this priority, why can't we have user account setting value take higher precedence?